### ISSUE TITLE
Queue needs bounds on its Send/Sync traits

### ISSUE THREAD

**Issue by ammaraskar** [Xudong-Huang/may#88]
Currently, it is possible to use `mpsc_list::Queue` and `mpsc_list_v1::Queue` to send across types that aren't safe to use in other threads. Due to the guarantee about there only being one popper, I think it should be Send/Sync when `T` is `Send`. Here's a demonstration with a data race using Cells:

```rust
#![forbid(unsafe_code)]

use may_queue::mpsc_list::Queue;

use std::cell::Cell;
use crossbeam_utils::thread;

#[derive(Debug, Clone, Copy)]
enum RefOrInt<'a> {
    Ref(&'a u64),
    Int(u64),
}
static SOME_INT: u64 = 123;

fn main() {
    let cell = Cell::new(RefOrInt::Ref(&SOME_INT));
    let queue = Queue::new();
    queue.push(&cell);

    thread::scope(|s| {
        s.spawn(move |_| {
            let smuggled_cell = queue.pop().unwrap();

            loop {
                // Repeatedly write Ref(&addr) and Int(0xdeadbeef) into the cell.
                smuggled_cell.set(RefOrInt::Ref(&SOME_INT));
                smuggled_cell.set(RefOrInt::Int(0xdeadbeef));
            }
        });

        loop {
            if let RefOrInt::Ref(addr) = cell.get() {
                // Hope that between the time we pattern match the object as a
                // `Ref`, it gets written to by the other thread.
                if addr as *const u64 == &SOME_INT as *const u64 {
                    continue;
                }

                // Due to the data race, obtaining Ref(0xdeadbeef) is possible
                println!("Pointer is now: {:p}", addr);
                println!("Dereferencing addr will now segfault: {}", *addr);
            }
        }
    });
}
```

which outputs:
```
Pointer is now: 0xdeadbeef

Return Code: -11 (SIGSEGV)
```

(Issue found by Rust group at @sslab-gatech)

**Comment by Xudong-Huang** [comment]
this should be fixed in release 0.3.19

**Comment by Xudong-Huang** [comment]
this should be fixed in release 0.3.19
